### PR TITLE
Added experimental support for Cry of Fear

### PR DESCRIPTION
**Pull request by SmileyAG**

### CODE REVIEW

File: BunnymodXT/cvars.cpp [hunk 1]
```diff
@@ -77,6 +78,7 @@ namespace CVars
 	CVarWrapper bxt_fire_on_button_target("bxt_fire_on_button_target", "");
 	CVarWrapper bxt_fire_on_button_command("bxt_fire_on_button_command", "");
 	CVarWrapper bxt_remove_stamina("bxt_remove_stamina", "0");
+	CVarWrapper bxt_cof_enable_ducktap("bxt_cof_enable_ducktap", "0");
```
Comment (YaLTeR) [inline BunnymodXT/cvars.cpp:81]: Why is this bxt_cof but the other cvar is bxt_..._in_cof?

File: BunnymodXT/hud_custom.cpp [hunk 2]
```diff
@@ -424,7 +424,7 @@ namespace CustomHud
 	void GetAccurateInfo()
 	{
 		receivedAccurateInfo = HwDLL::GetInstance().TryGettingAccurateInfo(player.origin, player.velocity, player.health, player.armorvalue, player.waterlevel, player.stamina);
-		HwDLL::GetInstance().GetViewangles(player.viewangles);
+		ClientDLL::GetInstance().pEngfuncs->GetViewAngles(player.viewangles);
```
Comment (YaLTeR) [inline BunnymodXT/hud_custom.cpp:427]: Doesn't this change replace an engine dependency with a client dependency? We don't want that because there's much more variety in clients than in the engine.

File: BunnymodXT/hud_custom.cpp [hunk 3]
```diff
@@ -1462,10 +1462,19 @@ namespace CustomHud
...
+				if (frame_bulk_selected)
+					out << frame_bulk_status.stamina;
+				else
+					out << player.stamina;
+			} else {
+				void* classPtr = (*hw.sv_player)->v.pContainingEntity->pvPrivateData;
+				uintptr_t thisAddr = reinterpret_cast<uintptr_t>(classPtr);
+				ptrdiff_t offm_fStamina = 0x21F0;
+				float* m_fStamina = reinterpret_cast<float*>(thisAddr + offm_fStamina);
+				out << *m_fStamina;
```
Comment (YaLTeR) [inline BunnymodXT/hud_custom.cpp:1476]: This should not be here, it should be wherever `player.stamina` is set.

File: BunnymodXT/modules/ClientDLL.cpp [hunk 4]
```diff
@@ -814,7 +820,7 @@ void ClientDLL::RegisterCVarsAndCommands()
 
 	#define REG(cvar) HwDLL::GetInstance().RegisterCVar(CVars::cvar)
 
-	if (ORIG_PM_Jump)
+	if (ORIG_PM_Jump && !HwDLL::GetInstance().is_cof)
```
Comment (YaLTeR) [inline BunnymodXT/modules/ClientDLL.cpp:823]: Why?

File: BunnymodXT/modules/ClientDLL.cpp [hunk 5]
```diff
@@ -1319,10 +1325,18 @@ HOOK_DEF_0(ClientDLL, void, __cdecl, HUD_Init)
...
 	{
 		orig_righthand_not_found = true;
-		HwDLL::GetInstance().RegisterCVar(CVars::cl_righthand);
+		hw.RegisterCVar(CVars::cl_righthand);
+	}
+
+	if (hw.is_cof)
+	{
+		hw.ORIG_Cmd_AddCommand("notarget", hw.ORIG_Host_Notarget_f);
+		hw.ORIG_Cmd_AddCommand("noclip", hw.ORIG_Host_Noclip_f);
```
Comment (YaLTeR) [inline BunnymodXT/modules/ClientDLL.cpp:1339]: Why is this in `ClientDLL::HUD_Init` and not in `HwDLL::RegisterCVarsAndCommandsIfNeeded`?

File: BunnymodXT/modules/ServerDLL.cpp [hunk 6]
```diff
@@ -1245,7 +1296,7 @@ void ServerDLL::RegisterCVarsAndCommands()
 
 	#define REG(cvar) HwDLL::GetInstance().RegisterCVar(CVars::cvar)
 	REG(bxt_timer_autostop);
-	if (ORIG_PM_Jump) {
+	if (ORIG_PM_Jump && !HwDLL::GetInstance().is_cof) {
```
Comment (YaLTeR) [inline BunnymodXT/modules/ServerDLL.cpp:1299]: Why?

File: BunnymodXT/modules/ServerDLL.cpp [hunk 7]
```diff
@@ -1260,7 +1311,7 @@ void ServerDLL::RegisterCVarsAndCommands()
 		REG(bxt_fire_on_stuck);
 	if (ORIG_CTriggerSave__SaveTouch || ORIG_CTriggerSave__SaveTouch_Linux)
 		REG(bxt_disable_autosave);
-	if (pCS_Stamina_Value)
+	if (pCS_Stamina_Value || HwDLL::GetInstance().is_cof)
```
Comment (YaLTeR) [inline BunnymodXT/modules/ServerDLL.cpp:1314]: Why?

File: BunnymodXT/modules/ServerDLL.hpp [hunk 8]
```diff
@@ -96,7 +97,7 @@ class ServerDLL : public IHookableDirFilter
 	void SetStamina(bool makeItZero);
 
 private:
-	ServerDLL() : IHookableDirFilter({ L"dlls" }) {};
+	ServerDLL() : IHookableDirFilter({ L"dlls", L"cl_dlls"}) {};
```
Comment (YaLTeR) [inline BunnymodXT/modules/ServerDLL.hpp:100]: Please add a CoF comment here too.

**Review by YaLTeR — CHANGES_REQUESTED** [review]

File: BunnymodXT/modules/ServerDLL.cpp (hunk 7 above)
Comment (SmileyAG) [reply BunnymodXT/modules/ServerDLL.cpp:1314]: For register cvar to disabling stamina limit in CoF, why it surprises you?

File: BunnymodXT/cvars.cpp (hunk 1 above)
Comment (SmileyAG) [reply BunnymodXT/cvars.cpp:81]: I'll rename `bxt_disable_save_lock_in_cof` to `bxt_cof_disable_save_lock`, thanks for point.

File: BunnymodXT/modules/ServerDLL.cpp (hunk 6 above)
Comment (SmileyAG) [reply BunnymodXT/modules/ServerDLL.cpp:1299]: `bxt_autojump` doesn't take stamina in that game after you made 1st jump in bhop sequence, while `+bxt_tas_autojump` and normal jumps does, so I think it illegal one cuz probably there is some button check in server that prevents to take stamina or smth like that, not inspected much for that question.

Hoboken suggested before to take off stamina in `bxt_autojump` code (since we already know where that memory value is placed), but that idea seems to me scuffed tbh, I rather ask runners use `+bxt_tas_autojump` and that's all.

File: BunnymodXT/modules/ClientDLL.cpp (hunk 4 above)
Comment (SmileyAG) [reply BunnymodXT/modules/ClientDLL.cpp:823]: https://github.com/YaLTeR/BunnymodXT/pull/333#discussion_r985079875

File: BunnymodXT/hud_custom.cpp (hunk 2 above)
Comment (hobokenn) [reply BunnymodXT/hud_custom.cpp:427]: since https://github.com/YaLTeR/BunnymodXT/commit/8917622040397e3eece04b6cbfd4dbe88fbc6cc9 it's an engine dependency, i tried to explain it a bit in my commit https://github.com/SmileyAG/BunnymodXT/pull/9/commits/7e7b42cea3b2104a33cab59804b0f94ad91eb761, but smiley misunderstood me and squashed my pr

File: BunnymodXT/modules/ServerDLL.cpp (hunk 6 above)
Comment (YaLTeR) [reply BunnymodXT/modules/ServerDLL.cpp:1299]: Ic

I think ideally +bxt_tas_autojump should start handling all cases it currently doesn't handle (see the OpenAG autojump) and then we can remove bxt_autojump altogether

File: BunnymodXT/hud_custom.cpp (hunk 3 above)
Comment (SmileyAG) [reply BunnymodXT/hud_custom.cpp:1476]: If I'm moving it to `SetStamina` custom function (that calls at end of **Cbuf_Execute**) it would cause a game crash state-like.

File: BunnymodXT/modules/ClientDLL.cpp (hunk 5 above)
Comment (SmileyAG) [reply BunnymodXT/modules/ClientDLL.cpp:1339]: If I do in `HwDLL::RegisterCVarsAndCommandsIfNeeded` it outputs **Cmd_AddCommand after host_initialized** error

Comment (YaLTeR) [reply BunnymodXT/modules/ClientDLL.cpp:1339]: What do you mean, we add a ton of commands there just fine?

File: BunnymodXT/modules/ServerDLL.cpp (hunk 6 above)
Comment (YaLTeR) [reply BunnymodXT/modules/ServerDLL.cpp:1301]: Could you add a comment to both here and the other place that bxt_autojump is disabled in CoF because it's extra unlegit?

File: BunnymodXT/modules/ServerDLL.cpp (hunk 7 above)
Comment (YaLTeR) [reply BunnymodXT/modules/ServerDLL.cpp:1316]: I didn't know that CoF had stamina.

File: BunnymodXT/hud_custom.cpp (hunk 3 above)
Comment (YaLTeR) [reply BunnymodXT/hud_custom.cpp:1476]: What is `SetStamina`? I'm talking about `TryGettingAccurateInfo` where `player.stamina` is set.

File: BunnymodXT/modules/HwDLL.cpp [hunk 9]
```diff
@@ -3863,6 +4019,12 @@ void HwDLL::RegisterCVarsAndCommandsIfNeeded()
 	using CmdWrapper::Handler;
 	typedef CmdWrapper::CmdWrapper<CmdFuncs> wrapper;
 
+	if (is_cof)
+	{
+		ServerDLL::GetInstance().pEngfuncs->pfnAddServerCommand("noclip", ORIG_Host_Noclip_f);
+		ServerDLL::GetInstance().pEngfuncs->pfnAddServerCommand("notarget", ORIG_Host_Notarget_f);
```
Comment (YaLTeR) [inline BunnymodXT/modules/HwDLL.cpp:4025]: No, if it is in `HwDLL`, it should use an engine command-adding function... The server isn't even loaded at this point

Comment (SmileyAG) [reply BunnymodXT/modules/HwDLL.cpp:4025]: 1) `pfnAddServerCommand` is not a server function, it's `Cmd_AddGameCommand` in engine.
2) That **pEngfuncs** table got hooked in engine too, find `auto fLoadThisDll = FindAsync(` line in **HwDLL.cpp**

TLDR; not need to worries at all.

**Comment by SmileyAG** [comment]
@YaLTeR I think I'm done with BXT supporting to CoF for now, we got everything that needs for RTA runs or segmenting, so free to merge or review.

**Comment by YaLTeR** [comment]
Thanks!
